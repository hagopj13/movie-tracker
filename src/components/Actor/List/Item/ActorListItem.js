// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import ImageIcon from '@material-ui/icons/Image';

import configSelectors from 'store/config/config.selectors';
import type { Actor } from 'types';

type Props = {
  actor: Actor,
  fullProfilePath: string,
};

const useStyles = makeStyles((theme) => ({
  root: {
    width: 155,
    minWidth: 155,
  },
  cardContent: {
    padding: theme.spacing(2),
    '&:last-child': {
      paddingBottom: theme.spacing(2),
    },
  },
  profileMedia: {
    width: 155,
    height: 235,
  },
  brokenMediaContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: theme.palette.grey[200],
    height: 235,
  },
}));

export const ActorListItem = (props: Props) => {
  const { actor, fullProfilePath } = props;

  const classes = useStyles();

  const isImageFound = actor.profilePath && fullProfilePath;

  const renderMedia = () => {
    return isImageFound ? (
      <CardMedia className={classes.profileMedia} image={fullProfilePath} />
    ) : (
      <div className={classes.brokenMediaContainer}>
        <ImageIcon fontSize="large" />
      </div>
    );
  };

  return (
    <Card className={classes.root} elevation={3}>
      {renderMedia()}
      <CardContent className={classes.cardContent}>
        <Typography variant="h6">{actor.name}</Typography>
        <Typography>{actor.character}</Typography>
      </CardContent>
    </Card>
  );
};

const mapStateToProps = createStructuredSelector({
  fullProfilePath: (state, ownProps: Props) =>
    configSelectors.createProfileFullPathSelector(ownProps.actor.profilePath)(state),
});

export default connect(mapStateToProps)(ActorListItem);
