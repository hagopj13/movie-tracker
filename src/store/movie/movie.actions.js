// @flow
import type { Movie, MovieUserState } from 'types';

import MovieActionTypes from './movie.types';

const fetchMovie = (id: string) => ({
  type: MovieActionTypes.FETCH_MOVIE,
  payload: { id },
});

const fetchMovieStart = () => ({
  type: MovieActionTypes.FETCH_MOVIE_START,
});

const fetchMovieSuccess = (movie: Movie, movieUserState: MovieUserState) => ({
  type: MovieActionTypes.FETCH_MOVIE_SUCCESS,
  payload: { movie, movieUserState },
});

const fetchMovieFailure = (error: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_FAILURE,
  payload: { error },
});

const fetchMovieUserState = (id: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE,
  payload: { id },
});

const fetchMovieUserStateStart = () => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_START,
});

const fetchMovieUserStateSuccess = (movieUserState: MovieUserState) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_SUCCESS,
  payload: { movieUserState },
});

const fetchMovieUserStateFailure = (error: string) => ({
  type: MovieActionTypes.FETCH_MOVIE_USER_STATE_FAILURE,
  payload: { error },
});

const setIsMovieFavorite = (id: number, isFavorite: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE,
  payload: { id, isFavorite },
});

const setIsMovieFavoriteStart = () => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_START,
});

const setIsMovieFavoriteSuccess = (id: number, isFavorite: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_SUCCESS,
  payload: { id, isFavorite },
});

const setIsMovieFavoriteFailure = (error: string) => ({
  type: MovieActionTypes.SET_IS_MOVIE_FAVORITE_FAILURE,
  payload: { error },
});

const setIsMovieInWatchlist = (id: string, isInWatchlist: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST,
  payload: { id, isInWatchlist },
});

const setIsMovieInWatchlistStart = () => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_START,
});

const setIsMovieInWatchlistSuccess = (id: number, isInWatchlist: boolean) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_SUCCESS,
  payload: { id, isInWatchlist },
});

const setIsMovieInWatchlistFailure = (error: string) => ({
  type: MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_FAILURE,
  payload: { error },
});

const rateMovie = (id: number, rating: number | null) => ({
  type: MovieActionTypes.RATE_MOVIE,
  payload: { id, rating },
});

const rateMovieStart = () => ({
  type: MovieActionTypes.RATE_MOVIE_START,
});

const rateMovieSuccess = (id: number, rating: number | null) => ({
  type: MovieActionTypes.RATE_MOVIE_SUCCESS,
  payload: { id, rating },
});

const rateMovieFailure = (error: string) => ({
  type: MovieActionTypes.RATE_MOVIE_FAILURE,
  payload: { error },
});

export default {
  fetchMovie,
  fetchMovieStart,
  fetchMovieSuccess,
  fetchMovieFailure,
  fetchMovieUserState,
  fetchMovieUserStateStart,
  fetchMovieUserStateSuccess,
  fetchMovieUserStateFailure,
  setIsMovieFavorite,
  setIsMovieFavoriteStart,
  setIsMovieFavoriteSuccess,
  setIsMovieFavoriteFailure,
  setIsMovieInWatchlist,
  setIsMovieInWatchlistStart,
  setIsMovieInWatchlistSuccess,
  setIsMovieInWatchlistFailure,
  rateMovie,
  rateMovieStart,
  rateMovieSuccess,
  rateMovieFailure,
};
