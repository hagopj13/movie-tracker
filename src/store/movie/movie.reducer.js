// @flow
import { handleActions } from 'redux-actions';

import AuthActionTypes from 'store/auth/auth.types';
import type { Movie, MovieUserState } from 'types';

import MovieActionTypes from './movie.types';

type State = {
  details: Movie,
  userState: MovieUserState,
};

export const initialState: State = {
  details: {
    id: null,
    title: '',
    releaseDate: '',
    voteAverage: 0,
    voteCount: 0,
    posterPath: '',
    backdropPath: '',
    tagline: '',
    overview: '',
    runtime: 0,
    status: '',
    genres: [],
    actors: [],
    reviews: [],
    recommendations: [],
  },
  userState: {
    isFavorite: false,
    isInWatchlist: false,
    rating: null,
  },
};

const fetchMovieStart = () => initialState;

const fetchMovieSuccess = (state: State, action) => ({
  ...state,
  details: action.payload.movie,
  userState: action.payload.movieUserState,
});

const fetchMovieUserStateSuccess = (state: State, action) => ({
  ...state,
  userState: action.payload.movieUserState,
});

const setIsMovieFavoriteSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        isFavorite: action.payload.isFavorite,
      },
    };
  }
  return state;
};

const setIsMovieInWatchlistSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        isInWatchlist: action.payload.isInWatchlist,
      },
    };
  }
  return state;
};

const rateMovieSuccess = (state: State, action) => {
  if (action.payload.id === state.details?.id) {
    return {
      ...state,
      userState: {
        ...state.userState,
        rating: action.payload.rating,
      },
    };
  }
  return state;
};

const logoutSuccess = (state: State) => ({
  ...state,
  userState: initialState.userState,
});

const movieActionHandler = {
  [MovieActionTypes.FETCH_MOVIE_START]: fetchMovieStart,
  [MovieActionTypes.FETCH_MOVIE_SUCCESS]: fetchMovieSuccess,
  [MovieActionTypes.FETCH_MOVIE_USER_STATE_SUCCESS]: fetchMovieUserStateSuccess,
  [MovieActionTypes.SET_IS_MOVIE_FAVORITE_SUCCESS]: setIsMovieFavoriteSuccess,
  [MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST_SUCCESS]: setIsMovieInWatchlistSuccess,
  [MovieActionTypes.RATE_MOVIE_SUCCESS]: rateMovieSuccess,
  [AuthActionTypes.LOGOUT_SUCCESS]: logoutSuccess,
};

export default handleActions(movieActionHandler, initialState);
