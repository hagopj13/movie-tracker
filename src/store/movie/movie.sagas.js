import { all, call, takeLatest, put, select } from 'redux-saga/effects';

import * as api from 'api/tmdb';
import authSelectors from 'store/auth/auth.selectors';
import { convertResponseToMovie, convertResponseToMovieUserState } from 'api/tmdb/utils';

import MovieActionTypes from './movie.types';
import movieActions from './movie.actions';

export function* fetchMovie({ payload: { id } }) {
  yield put(movieActions.fetchMovieStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    const { data } = yield call(api.getMovieDetails, id, sessionId);
    const movie = yield call(convertResponseToMovie, data);
    const movieUserState = yield call(convertResponseToMovieUserState, data.account_states);
    yield put(movieActions.fetchMovieSuccess(movie, movieUserState));
  } catch (error) {
    yield put(movieActions.fetchMovieFailure(error.status_message));
  }
}

export function* fetchMovieUserState({ payload: { id } }) {
  yield put(movieActions.fetchMovieUserStateStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    const { data } = yield call(api.getMovieUserState, id, sessionId);
    const movieUserState = yield call(convertResponseToMovieUserState, data);
    yield put(movieActions.fetchMovieUserStateSuccess(movieUserState));
  } catch (error) {
    yield put(movieActions.fetchMovieUserStateFailure(error.status_message));
  }
}

export function* setIsMovieFavorite({ payload: { id, isFavorite } }) {
  yield put(movieActions.setIsMovieFavoriteStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  const accountId = yield select(authSelectors.selectAccountId);
  try {
    yield call(api.setIsMovieFavorite, id, isFavorite, sessionId, accountId);
    yield put(movieActions.setIsMovieFavoriteSuccess(id, isFavorite));
  } catch (error) {
    yield put(movieActions.setIsMovieFavoriteFailure(error.status_message));
  }
}

export function* setIsMovieInWatchlist({ payload: { id, isInWatchlist } }) {
  yield put(movieActions.setIsMovieInWatchlistStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  const accountId = yield select(authSelectors.selectAccountId);
  try {
    yield call(api.setIsMovieInWatchlist, id, isInWatchlist, sessionId, accountId);
    yield put(movieActions.setIsMovieInWatchlistSuccess(id, isInWatchlist));
  } catch (error) {
    yield put(movieActions.setIsMovieInWatchlistFailure(error.status_message));
  }
}

export function* rateMovie({ payload: { id, rating } }) {
  yield put(movieActions.rateMovieStart());
  const sessionId = yield select(authSelectors.selectSessionId);
  try {
    if (rating) {
      yield call(api.rateMovie, id, rating * 2, sessionId);
    } else {
      yield call(api.deleteMovieRating, id, sessionId);
    }
    yield put(movieActions.rateMovieSuccess(id, rating));
  } catch (error) {
    yield put(movieActions.rateMovieFailure(error.status_message));
  }
}

export function* onFetchMovie() {
  yield takeLatest(MovieActionTypes.FETCH_MOVIE, fetchMovie);
}

export function* onFetchMovieUserState() {
  yield takeLatest(MovieActionTypes.FETCH_MOVIE_USER_STATE, fetchMovieUserState);
}

export function* onSetIsMovieFavorite() {
  yield takeLatest(MovieActionTypes.SET_IS_MOVIE_FAVORITE, setIsMovieFavorite);
}

export function* onSetIsMovieInWatchlist() {
  yield takeLatest(MovieActionTypes.SET_IS_MOVIE_IN_WATCHLIST, setIsMovieInWatchlist);
}

export function* onRateMovie() {
  yield takeLatest(MovieActionTypes.RATE_MOVIE, rateMovie);
}

export default function* movieSagas() {
  yield all([
    call(onFetchMovie),
    call(onFetchMovieUserState),
    call(onSetIsMovieFavorite),
    call(onSetIsMovieInWatchlist),
    call(onRateMovie),
  ]);
}
