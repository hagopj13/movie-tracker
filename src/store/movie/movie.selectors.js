// @flow
import { createSelector } from 'reselect';

const selectMovie = (state) => state.movie;

const selectMovieDetails = createSelector([selectMovie], (movie) => movie.details);

const selectMovieUserState = createSelector([selectMovie], (movie) => movie.userState);

export default {
  selectMovieDetails,
  selectMovieUserState,
};
