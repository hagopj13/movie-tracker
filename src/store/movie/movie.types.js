export default {
  FETCH_MOVIE: 'FETCH_MOVIE',
  FETCH_MOVIE_START: 'FETCH_MOVIE_START',
  FETCH_MOVIE_SUCCESS: 'FETCH_MOVIE_SUCCESS',
  FETCH_MOVIE_FAILURE: 'FETCH_MOVIE_FAILURE',
  FETCH_MOVIE_USER_STATE: 'FETCH_MOVIE_USER_STATE',
  FETCH_MOVIE_USER_STATE_START: 'FETCH_MOVIE_USER_STATE_START',
  FETCH_MOVIE_USER_STATE_SUCCESS: 'FETCH_MOVIE_USER_STATE_SUCCESS',
  FETCH_MOVIE_USER_STATE_FAILURE: 'FETCH_MOVIE_USER_STATE_FAILURE',
  SET_IS_MOVIE_FAVORITE: 'SET_IS_MOVIE_FAVORITE',
  SET_IS_MOVIE_FAVORITE_START: 'SET_IS_MOVIE_FAVORITE_START',
  SET_IS_MOVIE_FAVORITE_SUCCESS: 'SET_IS_MOVIE_FAVORITE_SUCCESS',
  SET_IS_MOVIE_FAVORITE_FAILURE: 'SET_IS_MOVIE_FAVORITE_FAILURE',
  SET_IS_MOVIE_IN_WATCHLIST: 'SET_IS_MOVIE_IN_WATCHLIST',
  SET_IS_MOVIE_IN_WATCHLIST_START: 'SET_IS_MOVIE_IN_WATCHLIST_START',
  SET_IS_MOVIE_IN_WATCHLIST_SUCCESS: 'SET_IS_MOVIE_IN_WATCHLIST_SUCCESS',
  SET_IS_MOVIE_IN_WATCHLIST_FAILURE: 'SET_IS_MOVIE_IN_WATCHLIST_FAILURE',
  RATE_MOVIE: 'RATE_MOVIE',
  RATE_MOVIE_START: 'RATE_MOVIE_START',
  RATE_MOVIE_SUCCESS: 'RATE_MOVIE_SUCCESS',
  RATE_MOVIE_FAILURE: 'RATE_MOVIE_FAILURE',
};
