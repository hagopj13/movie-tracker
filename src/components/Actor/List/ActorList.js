// @flow
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';

import type { Actor } from 'types';

import ActorListItem from './Item/ActorListItem';

type Props = {
  actors: Actor[],
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    margin: theme.spacing(-2.25, 0, 0, -2.25),
    '& > *': {
      margin: theme.spacing(2.25, 0, 0, 2.25),
    },
    [theme.breakpoints.down('xs')]: {
      justifyContent: 'center',
    },
  },
  noActorsText: {
    fontWeight: theme.typography.fontWeightRegular,
  },
}));

export const ActorList = (props: Props) => {
  const { actors } = props;

  const classes = useStyles();

  if (actors.length === 0) {
    return (
      <Typography className={classes.noActorsText} variant="h6">
        No actors found
      </Typography>
    );
  }

  return (
    <div className={classes.root}>
      {actors.map((actor) => (
        <ActorListItem key={actor.id} actor={actor} />
      ))}
    </div>
  );
};

export default ActorList;
