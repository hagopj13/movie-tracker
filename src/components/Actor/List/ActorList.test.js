import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import { details as movieDetails } from 'store/fixtures/movie';

import { ActorList } from './ActorList';
import ActorListItem from './Item/ActorListItem';

describe('ActorList component', () => {
  let shallow;
  let actors;
  let wrapper;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    actors = movieDetails.actors;
    const mockProps = {
      actors,
    };
    wrapper = shallow(<ActorList {...mockProps} />);
  });

  it('should render the ActorList component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the actors as a list of ActorListItem', () => {
    expect(wrapper.find(ActorListItem)).toHaveLength(actors.length);
  });

  it('should render a text stating there are no actors if the actors prop is an empty array', () => {
    const mockProps = {
      actors: [],
    };
    const newWrapper = shallow(<ActorList {...mockProps} />);
    expect(newWrapper.text()).toBe('No actors found');
  });
});
