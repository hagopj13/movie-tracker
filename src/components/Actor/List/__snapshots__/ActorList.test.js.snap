// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActorList component should render the ActorList component correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <Connect(ActorListItem)
    actor={
      Object {
        "character": "Sherlock Holmes",
        "id": 3223,
        "name": "Robert Downey Jr.",
        "profilePath": "/1YjdSym1jTG7xjHSI0yGGWEsw5i.jpg",
      }
    }
    key="3223"
  />
  <Connect(ActorListItem)
    actor={
      Object {
        "character": "Dr. John Watson",
        "id": 9642,
        "name": "Jude Law",
        "profilePath": "/aQD9e9SPjxCdyqxyXKBmEhNgjtR.jpg",
      }
    }
    key="9642"
  />
  <Connect(ActorListItem)
    actor={
      Object {
        "character": "Professor James Moriarty",
        "id": 15440,
        "name": "Jared Harris",
        "profilePath": "/mGODbYpkR1NKBOV3rtTfZ7EWoIl.jpg",
      }
    }
    key="15440"
  />
</div>
`;
